File: cc/containers/hello_world_enclave_app/standalone_test.cc
/*
 * Copyright 2024 The Project Oak Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cstdint>

#include "absl/log/log.h"
#include "absl/status/status_matchers.h"
#include "absl/strings/substitute.h"
#include "cc/containers/hello_world_enclave_app/app_service.h"
#include "cc/containers/sdk/standalone/oak_standalone.h"
#include "cc/ffi/rust_bytes.h"
#include "cc/oak_session/client_session.h"
#include "gmock/gmock.h"
#include "grpcpp/server.h"
#include "grpcpp/server_builder.h"
#include "gtest/gtest.h"
#include "oak_containers/examples/hello_world/proto/hello_world.grpc.pb.h"

namespace oak::containers::sdk::standalone {

namespace {

using ::absl_testing::IsOk;
using ::absl_testing::IsOkAndHolds;
using ::oak::containers::example::EnclaveApplication;
using ::oak::containers::hello_world_enclave_app::EnclaveApplicationImpl;
using ::oak::session::v1::EndorsedEvidence;
using ::oak::session::v1::SessionRequest;
using ::oak::session::v1::SessionResponse;
using ::testing::Eq;

constexpr absl::string_view kApplicationConfig = "{}";

class HelloWorldStandaloneTest : public testing::Test {
  void SetUp() override {
    service_ = std::make_unique<EnclaveApplicationImpl>(kApplicationConfig);

    grpc::ServerBuilder builder;
    builder.AddListeningPort("[::]:8080", grpc::InsecureServerCredentials());
    builder.RegisterService(service_.get());
    server_ = builder.BuildAndStart();

    auto channel = server_->InProcessChannel({});
    stub_ = EnclaveApplication::NewStub(channel);
  }

 protected:
  std::unique_ptr<EnclaveApplicationImpl> service_;
  std::unique_ptr<grpc::Server> server_;
  std::unique_ptr<EnclaveApplication::Stub> stub_;
};

TEST_F(HelloWorldStandaloneTest, OakSessionReturnsResponse) {
  absl::StatusOr<std::unique_ptr<session::ClientSession>> session =
      session::ClientSession::Create(
          session::SessionConfigBuilder(session::AttestationType::kUnattested,
                                        session::HandshakeType::kNoiseNN)
              .Build());

  grpc::ClientContext context;
  std::unique_ptr<grpc::ClientReaderWriterInterface<
      session::v1::SessionRequest, session::v1::SessionResponse>>
      stream = stub_->OakSession(&context);

  // Handshake
  while (!(*session)->IsOpen()) {
    absl::StatusOr<std::optional<session::v1::SessionRequest>>
        outgoing_message = (*session)->GetOutgoingMessage();
    ASSERT_THAT(outgoing_message, IsOk());
    ASSERT_TRUE(stream->Write(**outgoing_message));

    session::v1::SessionResponse response;
    ASSERT_TRUE(stream->Read(&response));
    ASSERT_THAT((*session)->PutIncomingMessage(response), IsOk());
  }

  ASSERT_THAT((*session)->Write("Standalone Test"), IsOk());
  absl::StatusOr<std::optional<session::v1::SessionRequest>> outgoing_message =
      (*session)->GetOutgoingMessage();
  ASSERT_THAT(outgoing_message, IsOk());
  ASSERT_TRUE(stream->Write(**outgoing_message));

  session::v1::SessionResponse response;
  ASSERT_TRUE(stream->Read(&response));
  ASSERT_THAT((*session)->PutIncomingMessage(response), IsOk());
  absl::StatusOr<std::optional<ffi::RustBytes>> unencrypted_response =
      (*session)->ReadToRustBytes();

  ASSERT_THAT(unencrypted_response,
              IsOkAndHolds(std::optional(absl::Substitute(
                  "Hello from the enclave, Standalone Test! Btw, the "
                  "app has a config with a length of $0 bytes.",
                  kApplicationConfig.size()))));
}

}  // namespace

}  // namespace oak::containers::sdk::standalone

File: oak_containers/examples/hello_world/enclave_app/tests/standalone_test.rs
//
// Copyright 2024 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::net::{IpAddr, Ipv4Addr, SocketAddr};

use anyhow::{Context, Result};
use futures::channel::mpsc;
use oak_hello_world_proto::oak::containers::example::enclave_application_client::EnclaveApplicationClient;
use oak_session::{
    Session,
    attestation::AttestationType,
    channel::{SessionChannel, SessionInitializer},
    config::SessionConfig,
    handshake::HandshakeType,
};
use tokio::net::TcpListener;
use tonic::transport::Channel;

const APPLICATION_CONFIG: &[u8] = b"fake_config";

async fn start_server() -> Result<(SocketAddr, tokio::task::JoinHandle<Result<()>>)> {
    let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0);
    let listener = TcpListener::bind(addr).await?;
    let addr = listener.local_addr()?;

    Ok((
        addr,
        tokio::spawn(oak_containers_examples_hello_world_enclave_app::app_service::create(
            listener,
            oak_containers_examples_hello_world_enclave_app::app::HelloWorldApplicationHandler {
                application_config: APPLICATION_CONFIG.to_vec(),
            },
        )),
    ))
}

#[tokio::test]
async fn test_noise() {
    // Start server
    let (addr, _join_handle) = start_server().await.unwrap();

    let url = format!("http://{addr}");

    println!("Connecting to test server on {}", url);

    let channel = Channel::from_shared(url)
        .context("couldn't create gRPC channel")
        .unwrap()
        .connect()
        .await
        .context("couldn't connect via gRPC channel")
        .unwrap();

    let mut client = EnclaveApplicationClient::new(channel);

    let (mut tx, rx) = mpsc::channel(10);

    let mut response_stream =
        client.oak_session(rx).await.expect("couldn't send stream request").into_inner();

    // We don't have a noise client impl yet, so we need to manage the session
    // manually.
    let mut client_session = oak_session::ClientSession::create(
        SessionConfig::builder(AttestationType::Unattested, HandshakeType::NoiseNN).build(),
    )
    .expect("could not create client session");

    while !client_session.is_open() {
        let request = client_session.next_init_message().expect("expected client init message");
        tx.try_send(request).expect("failed to send to server");
        if !client_session.is_open() {
            let response = response_stream
                .message()
                .await
                .expect("expected a response")
                .expect("response was failure");
            client_session.handle_init_message(response).expect("failed to handle init response");
        }
    }

    let request = client_session.encrypt(b"standalone user").expect("failed to send request");
    tx.try_send(request).expect("failed to send request");

    let result = response_stream
        .message()
        .await
        .expect("no response ready")
        .expect("failed to get response");
    let result = client_session.decrypt(result).expect("failed to decrypt result");
    let app_config_len = APPLICATION_CONFIG.len();
    assert_eq!(
        String::from_utf8(result).unwrap(),
        format!(
            "Hello from the enclave, standalone user! Btw, the app has a config with a length of {app_config_len} bytes."
        ),
    );
}
